shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm

```diff
@@ -255,7 +253,6 @@ static void ReplaceSemanticsObject(SemanticsObject* oldObject,
   FML_DCHECK(oldObject.node.id == newObject.uid);
   NSNumber* nodeId = @(oldObject.node.id);
   NSUInteger positionInChildlist = [oldObject.parent.children indexOfObject:oldObject];
-  [[oldObject retain] autorelease];
```

Comment:
This was odd practice here before. Did we add this line to fix a crash before? If so, this line didn't 100% fix that crash since it's still used in the previous 3 lines.

Reply:
Ah, yes! I noticed this line too and was going to leave a comment here referencing the commit here: https://github.com/flutter/engine/pull/31351

This bumps the refcount to ensure that `oldObject` isn't released if the refcount is decremented in this method itself. It's a relatively common pattern in non-ARC tree-related Obj-C code.

I didn't mentally work through the gymnastics of how `oldObject`'s refcount could be indirectly knocked down one in this method, but we are manipulating the children of our parent node here, which seems like the exact sort of case where that can happen.

Reply:
Let's try mentally working through it. What if `oldObject == newObject`? Then we get the position the list of `oldObject` into `positionInChildlist`, then we remove `oldObject` from `objects`, and that releases `oldObject` and now... if we didn't retain `oldObject`, `oldObject` is dead, but we call `oldObject.parent ...` on the next line. Boom.

Reply:
Ah in fact there's a [comment](https://github.com/flutter/engine/pull/31351#issuecomment-1034099886) that @stuartmorgan left pointing out this exact scenario if you scroll down far enough on the patch.

Reply:
It's 11 years old, but I think this is the talk I was thinking of that covers a simpler version of this scenario, though I swear there was a talk where they went through more complex scenarios like trees.

https://www.youtube.com/watch?v=ANlpDUIF-1o#t=9m19s

Reply:
I think the TL;DR here is that this line is an excellent argument in favour of ARC migration.

Reply:
> This is the sort of case that motivated the creation of ARC

Yes, I think it's a great sign of how successful ARC was that there are now a lot of people whose reaction to this line is "that's weird" rather than immediately knowing exactly why it's there and what it's doing. Everyone learning this pattern (usually the hard way) is definitely something we can do without 🙂 

> I think the TL;DR here is that this line is an excellent argument in favour of ARC migration.

Both the line and the discussion! The line because not having this kind of foot gun in the first place is clearly a safety win, but also the discussion because being able to correctly maintain non-ARC code is a shrinking/fading knowledge domain, so any non-ARC code we have will be an increasing liability over time.